### PR TITLE
Copy paste into same folder

### PR DESCRIPTION
**Pull request by alexanderKononov-sap**
Fix for #1523

### CODE REVIEW

**Comment by BeckerWdf** [comment]
@alexanderKononov-sap: Thanks for this PR and for signing the ECA.

**Comment by BeckerWdf** [comment]
I manually tested this in the project explorer. It looks good to me.
@vogella: Maybe you can test with your use-case as well.

File: bundles/org.eclipse.ui.navigator.resources/src/org/eclipse/ui/internal/navigator/resources/actions/PasteAction.java [hunk 1]
```diff
@@ -99,6 +107,22 @@ private IResource getTarget() {
...
+	private boolean selectionIsSameAsClipboardContent(List<? extends IResource> selectedResources,
+			List<? extends IResource> clipboardContent) {
+		if (clipboardContent != null && selectedResources != null && !selectedResources.isEmpty()) {
+			if (selectedResources.size() != clipboardContent.size()) {
+				return false;
+			}
+			return selectedResources.containsAll(clipboardContent);
+		}
+		return false;
+	}
```
Comment (fedejeanne) [inline bundles/org.eclipse.ui.navigator.resources/src/org/eclipse/ui/internal/navigator/resources/actions/PasteAction.java]: This method is an "equals" but it allows for different order of the items, so I think this more generic version is more appropriate:

```suggestion
	/**
	 * @param a The first collection
	 * @param b The second collection
	 * @return true if both collections aren't null and they contain the exact same
	 *         items, regardless of their order.
	 */
	private boolean areEqualsUnordered(Collection<? extends IResource> a, Collection<? extends IResource> b) {
		return b != null && a != null && !a.isEmpty() // they are not empty...
				&& a.size() == b.size() // ... and they have the same size
				&& a.containsAll(b); // ... and all elements of A are in B
	}
```

If performance ever becomes an issue (please test for it) then you could change the implementation so it first sorts both lists (or a copy of the list if you don't want to change the original lists) and then calls `equals` on them.

File: bundles/org.eclipse.ui.navigator.resources/src/org/eclipse/ui/internal/navigator/resources/actions/PasteAction.java [hunk 2]
```diff
@@ -146,17 +170,25 @@ public void run() {
...
 	 * Returns the container to hold the pasted resources.
+	 *
+	 * @param clipboardContent current content of the clipboard
 	 */
-	private IContainer getContainer() {
+	private IContainer getContainer(IResource[] clipboardContent) {
 		List<? extends IResource> selection = getSelectedResources();
+
+		// selection is copied to itself => copy to parent
+		if (clipboardContent != null && selectionIsSameAsClipboardContent(selection, Arrays.asList(clipboardContent))) {
```
Comment (fedejeanne) [inline bundles/org.eclipse.ui.navigator.resources/src/org/eclipse/ui/internal/navigator/resources/actions/PasteAction.java]:
```suggestion
		if (clipboardContent != null && areEqualsUnordered(selection, Arrays.asList(clipboardContent))) {
```

**Review by fedejeanne — COMMENTED** [review]
Hello @alexanderKononov-sap , thank you for contributing.

The code looks good, I just have a minor remark on the name and the implementation of the new method.

DISCLAIMER: I haven't tested the code yet, I just focused on the coding practices :-)

**Comment by vogella** [comment]
@alexanderKononov-sap thanks, code works fine. I tested in project explorer.

Please combine your two commits and write a good commit message, something like:

-----------
Enhance copy of a folder in project explorer.  Previous to fix fix you had to move to the top-level folder to paste a copied folder. This is not common in other tools.

Example:

FolderTop
-> Subfolder1
-> Subfolders99

Copy Subfolders99. Before this fix you had to move FolderTop to paste it, now you do not have to move the cursor.

Fixes: https://github.com/eclipse-platform/eclipse.platform.ui/issues/1523

-----------

The main source of documentation should be the Git history.

**Comment by vogella** [comment]
After this change and addressing the comment of @fedejeanne I plan to merge this change. Thanks again @alexanderKononov-sap for fixing this super ugly UX issue in the Eclipse IDE.

**Review by fedejeanne — CHANGES_REQUESTED** [review]
I just tested the new functionality and found an error.

Here's a screenshot of a simple java project before this PR (left) and after this PR (right, in red)

image

You can see that the _Project Explorer_ has a nested package called `myproject.copy2` under the package `myproject` after applying this PR, which is wrong. The _Package Explorer_ doesn't have this issue though.

**Steps to reproduce**
* Open the _Project Explorer_
* Copy and paste the package `myproject` into itself

~I also noticed while using the _Project explorer_ that copying + pasting using the keyboard doesn't always react instantly, it takes some tries to get it to react. I haven't been able to reproduce this always, it just happened some times. Maybe a performance issue with the new methods?~

Please analyze and fix.

EDIT: the issue with copying and pasting with the keyboard is probably my keyboard/environment, I'm experiencing the same issue in other apps. You can ignore it :-)

File: bundles/org.eclipse.ui.navigator.resources/src/org/eclipse/ui/internal/navigator/resources/actions/PasteAction.java (hunk 1 above)
Comment (alexanderKononov-sap) [reply bundles/org.eclipse.ui.navigator.resources/src/org/eclipse/ui/internal/navigator/resources/actions/PasteAction.java]: Thanks for the feedback, I'll change that

**Comment by alexanderKononov-sap** [comment]
> I just tested the new functionality and found an error.
> 
> Here's a screenshot of a simple java project before this PR (left) and after this PR (right, in red)
> 
> image
> 
> You can see that the _Project Explorer_ has a nested package called `myproject.copy2` under the package `myproject` after applying this PR, which is wrong. The _Package Explorer_ doesn't have this issue though.
> 
> **Steps to reproduce**
> 
> * Open the _Project Explorer_
> * Copy and paste the package `myproject` into itself
> 
> ~I also noticed while using the _Project explorer_ that copying + pasting using the keyboard doesn't always react instantly, it takes some tries to get it to react. I haven't been able to reproduce this always, it just happened some times. Maybe a performance issue with the new methods?~
> 
> Please analyze and fix.
> 
> EDIT: the issue with copying and pasting with the keyboard is probably my keyboard/environment, I'm experiencing the same issue in other apps. You can ignore it :-)

Hi @fedejeanne! Thank you for testing it!
Somehow I do not understand what is happening there...

I am able to reproduce the issue too, but I cannot explain it, maybe due to my lack of deep knowledge of Eclipse :)

I was under the impression that Java project tree is completely handled in JDT (even if it is shown in *project explorer*, it still behaves like *package explorer*). That is why I also did not test it that much. And I am still under that impression, although confused.

The thing is, I do see the difference in behavior, but when I test it, my new coding does not seem to be used! I set breakpoints in all the methods and nothing seems to be used. Can it be that JDT also made some changes and this is because of that?

Help! :)

--EDIT: just tested it again in target without my new changes and it behaves the same. So I am guessing that you have also tested in the IDE, not in target? Apparently there were some JDT changes in between... can you confirm?

**Comment by iloveeclipse** [comment]
Have you seen that in Package explorer we have `org.eclipse.jdt.internal.ui.refactoring.reorg.PasteAction` and in Project explorer `org.eclipse.ui.internal.navigator.resources.actions.PasteAction`?

File: bundles/org.eclipse.ui.navigator.resources/src/org/eclipse/ui/internal/navigator/resources/actions/PasteAction.java [hunk 3]
```diff
@@ -80,11 +81,18 @@ public PasteAction(Shell shell, Clipboard clipboard) {
...
+	 * @param clipboardContent current content of the clipboard
+	 *
 	 * @return the actual target of the paste action
 	 */
-	private IResource getTarget() {
+	private IResource getTarget(IResource[] clipboardContent) {
 		List<? extends IResource> selectedResources = getSelectedResources();
 
+		// selection is copied to itself => copy to parent
+		if (clipboardContent != null && selectionIsSameAsClipboardContent(selectedResources, Arrays.asList(clipboardContent))) {
```
Comment (fedejeanne) [inline bundles/org.eclipse.ui.navigator.resources/src/org/eclipse/ui/internal/navigator/resources/actions/PasteAction.java]:
```suggestion
		if (clipboardContent != null && areEqualsUnordered(selectedResources, Arrays.asList(clipboardContent))) {
```

**Review by fedejeanne — CHANGES_REQUESTED** [review]
> --EDIT: just tested it again in target without my new changes and it behaves the same. So I am guessing that you have also tested in the IDE, not in target? Apparently there were some JDT changes in between... can you confirm?

Sorry for the late response. You're right, I just tested in **I20240425-1800** and the problem is reproducible, so it was not introduced by this PR:

image

You forgot to change one call to `selectionIsSameAsClipboardContent` though. Also, I think some unwanted commits slipped in, please check that too and after that this PR is ✔️ .

You can ping me after the freeze period is finished and I can merge it.

**Comment by alexanderKononov-sap** [comment]
Sorry for the late reaction too, I am on vacation :) I hope it's fine now?

File: bundles/org.eclipse.ui.navigator.resources/src/org/eclipse/ui/internal/navigator/resources/actions/PasteAction.java (hunk 1 above)
Comment (BeckerWdf) [reply bundles/org.eclipse.ui.navigator.resources/src/org/eclipse/ui/internal/navigator/resources/actions/PasteAction.java]: @alexanderKononov-sap: Did you have a look on performance. Was the proposal "first sort and then call equals" not necesarry?

**Review by fedejeanne — APPROVED** [review]
> Sorry for the late reaction too, I am on vacation :) I hope it's fine now?

It's better, thank you. 

Just out of curiosity, have you looked into the other `PasteAction`s as commented before? https://github.com/eclipse-platform/eclipse.platform.ui/pull/1842#issuecomment-2075307072 . Is the new functionality also necessary there?

This PR is good to go. I checked the performance by copying a package with 100 classes into itself and I saw no performance degradation.

File: bundles/org.eclipse.ui.navigator.resources/src/org/eclipse/ui/internal/navigator/resources/actions/PasteAction.java (hunk 1 above)
Comment (fedejeanne) [reply bundles/org.eclipse.ui.navigator.resources/src/org/eclipse/ui/internal/navigator/resources/actions/PasteAction.java]: @BeckerWdf performance is still good, I checked by copying a folder with 100 classes into itself

**Comment by jukzi** [comment]
@fedejeanne the used commit message violate best practice, please update commit message next time to have a single line summary and link to the issue
image

**Pull request merged**